File: abc/abc158/abc158_d.cs
// https://atcoder.jp/contests/abc158/submissions/10646112

using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharp
{
    class Program
    {
        static void Main()
        {
            var S = Console.ReadLine();
            var Q = int.Parse(Console.ReadLine());

            var flag = true;
            var ll = new LinkedList<char>();
            ll.AddLast('*');

            for (int i = 0; i < Q; i++)
            {
                var input = Console.ReadLine();
                if (input[0] == '1')
                {
                    if (flag)
                    {
                        flag = false;
                    }
                    else
                    {
                        flag = true;
                    }
                    continue;
                }
                var q = input.Split().ToArray();
                if (flag)
                {
                    if (q[1] == "1")
                    {
                        ll.AddFirst(char.Parse(q[2]));
                    }
                    else
                    {
                        ll.AddLast(char.Parse(q[2]));
                    }
                }
                else
                {
                    if (q[1] == "1")
                    {
                        ll.AddLast(char.Parse(q[2]));
                    }
                    else
                    {
                        ll.AddFirst(char.Parse(q[2]));
                    }
                }
            }

            var str = new string(ll.ToArray()).Replace("*", S);
            Console.WriteLine(flag ? str : new string(str.Reverse().ToArray()));
        }
    }
}

File: abc/abc158/abc158_d.fs
// https://atcoder.jp/contests/abc158/submissions/11770667

open System.Collections.Generic
open System.Linq

let S = stdin.ReadLine()
let Q = stdin.ReadLine() |> int

let mutable flag = true
let ll = new LinkedList<char>()

'*'
|> ll.AddLast
|> ignore

for i in [ 0..Q - 1 ] do
    let input = stdin.ReadLine()
    if (input.[0] = '1') then flag <- not flag
    else
        let q = input.Split().ToArray()
        match flag with
        | true ->
            if (q.[1] = "1") then
                q.[2]
                |> char
                |> ll.AddFirst
                |> ignore
            else
                q.[2]
                |> char
                |> ll.AddLast
                |> ignore
        | false ->
            if (q.[1] = "1") then
                q.[2]
                |> char
                |> ll.AddLast
                |> ignore
            else
                q.[2]
                |> char
                |> ll.AddFirst
                |> ignore

let str = (new string(ll.ToArray())).Replace("*", S)

if flag then str
else new string(str.Reverse().ToArray())
|> printfn "%s"
